### PR TITLE
RAdam Update Rule

### PR DESCRIPTION
**Pull request by shubham1206agra**
Added RAdam Update Rule to Adam optimizers.
References:
- https://pytorch.org/docs/stable/generated/torch.optim.RAdam.html#torch.optim.RAdam
- https://arxiv.org/abs/1908.03265

### CODE REVIEW

File: include/ensmallen_bits/adam/radam_update.hpp [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+   public:
+    /**
+     * This constructor is called by the SGD Optimize() method before the start
+     * of the iteration update process.
+     *
+     * @param parent AdamUpdate object.
+     * @param rows Number of rows in the gradient matrix.
+     * @param cols Number of columns in the gradient matrix.
+     */
+    Policy(RAdamUpdate& parent, const size_t rows, const size_t cols) :
```
Comment (conradsnicta) [inline include/ensmallen_bits/adam/radam_update.hpp:92]: For clarity, suggest to use `n_rows` and `n_cols` instead of `rows` and `cols`.  
This will also make it consistent with Armadillo: http://arma.sourceforge.net/docs.html#attributes

File: include/ensmallen_bits/adam/radam_update.hpp [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    {
+      // Increment the iteration counter variable.
+      ++parent.iteration;
+
+      // And update the iterate.
+      m *= parent.beta1;
+      m += (1 - parent.beta1) * gradient;
+
+      v *= parent.beta2;
+      v += (1 - parent.beta2) * (gradient % gradient);
```
Comment (conradsnicta) [inline include/ensmallen_bits/adam/radam_update.hpp:118]: suggest to use `arma::square(gradient)` instead of `gradient % gradient`

File: tests/adam_test.cpp [hunk 3]
```diff
@@ -217,6 +217,15 @@ TEST_CASE("PadamLogisticRegressionTest", "[AdamTest]")
   LogisticRegressionFunctionTest(optimizer, 0.003, 0.006);
 }
 
+/**
+ * Run RAdam on logistic regression and make sure the results are acceptable.
+ */
+TEST_CASE("RAdamLogisticRegressionTest", "[AdamTest]")
+{
+  RAdam optimizer;
+  LogisticRegressionFunctionTest(optimizer, 0.003, 0.006);
```
Comment (conradsnicta) [inline tests/adam_test.cpp:226]: might be useful to provide a brief comment on why these settings (0.003, 0.006) are used.

File: include/ensmallen_bits/adam/radam_update.hpp (hunk 1 above)
Comment (zoq) [reply include/ensmallen_bits/adam/radam_update.hpp:92]: For ensmallen we basically followed the mlpack style guidelines where we use `cols, rows` but I can see that this creates some confusion, so I don't mind to adapt the armadillo style here.

File: include/ensmallen_bits/adam/radam_update.hpp (hunk 2 above)
Comment (zoq) [reply include/ensmallen_bits/adam/radam_update.hpp:118]: Ohh nice idea, we use `gradient % gradient` at some other places as well, so we should adapt it in another points as well.

File: include/ensmallen_bits/adam/radam_update.hpp [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+      const double biasCorrection1 = 1.0 - std::pow(parent.beta1,
+          parent.iteration);
+      const double biasCorrection2 = 1.0 - std::pow(parent.beta2,
+          parent.iteration);
+			const double pt = parent.pinf - 2.0 * parent.iteration * 
+					(1.0 - biasCorrection2) / biasCorrection2;
+			
+			if (pt > 5.0)
```
Comment (zoq) [inline include/ensmallen_bits/adam/radam_update.hpp:127]: The reference implementation uses `>= 5`, the paper uses `> 4`, so I would suggest to make this configurable and use the default from the reference implementation.

File: include/ensmallen_bits/adam/radam_update.hpp [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+      m += (1 - parent.beta1) * gradient;
+
+      v *= parent.beta2;
+      v += (1 - parent.beta2) * (gradient % gradient);
+
+      const double biasCorrection1 = 1.0 - std::pow(parent.beta1,
+          parent.iteration);
+      const double biasCorrection2 = 1.0 - std::pow(parent.beta2,
+          parent.iteration);
+			const double pt = parent.pinf - 2.0 * parent.iteration *
```
Comment (zoq) [inline include/ensmallen_bits/adam/radam_update.hpp:124]: Looks like the indentation is off in a couple of places, can you make sure it aligns with the rest of the codebase.

File: include/ensmallen_bits/adam/radam_update.hpp (hunk 1 above)
Comment (rcurtin) [reply include/ensmallen_bits/adam/radam_update.hpp:92]: I think we should stick with the mlpack style here, since ensmallen is part of the mlpack "umbrella".  It's good to have consistency across those projects (I know that this comes at the price of consistency with Armadillo).

**Review by rcurtin — COMMENTED** [review]
I think everyone else has already left substantive comments about the implementation, but one issue is, how will users find it?  Should we create a new optimizer called `Radam` and document it like the rest?  That seems like the best idea to me, personally.  Take a look at the implementation of, e.g., `Nadam`, and the associated documentation in `doc/optimizers.md` and `doc/function_types.md` for an example. :+1:

Also, don't forget to add a changelog entry. :)
